### ISSUE TITLE
More SSR Friendly Implementation

### ISSUE THREAD

**Issue by tatemz** [formaat-design/reshaped#352]
**Is your feature request related to a problem? Please describe.**
This is not a problem, but feedback. Components like `Accordion`, `Modal`, and `DropdownMenu` that require client-side Javascript to provider interactivity. Usually, the kind of user interactivity in these components is built to conditionally show or hide elements based on user interaction.

**Describe the solution you'd like**
Modern browsers are shipped with many a11y-friendly interactive UI elements that can be customized with CSS (and sometimes extended with JS). Where possible, I believe it would be ideal for components to use pure HTML + CSS solutions that do not require JS. This paradigm would require a shift in thinking in order to transform _stateful_ React components into _stateless_ ones. Building user interactivity without client-side Javascript oftentimes requires CSS "hacks and tricks" that can be tough to maintain. Abstracting complexity behind dev-friendly components is a noble goal.

I don't have a fully fledged prescription for Reshaped to adopt this paradigm, but I believe that by creating this issue and documenting this idea could spark healthy discussion around what is possible.

There are some elements within the Reshaped component library that re-implement existing browser functionality. Here is table showing possible mappings. I'm sure there are endless possibilities, but here are a few I notice off the top of my head

| Reshaped Component | Browser Functionality |
| -------- | ------- |
| `Accordion` | `<details />` |
| `Modal` | `<dialog />` |
| `Slider` | `<input type="range" />` [with styles](https://www.smashingmagazine.com/2021/12/create-custom-range-input-consistent-browsers/) |
| `Tabs` | [`:target` selectors](https://css-tricks.com/functional-css-tabs-revisited/)

Here is a Codepen that provides examples of ways to toggle visibility without JS.
https://codepen.io/tatemz/pen/NWQOmJW

One added benefit of using built-in semantic HTML tags is that a11y just works.

**Describe alternatives you've considered**

Trying to build components without client-side JS is a purists idea, and I know that it can come with tradeoffs. I'm confident Reshaped devs have thought through these tradeoffs before, and maybe there is a way to resurface some of these ideas. I'm a bit obsessive when it comes to making websites in that I want as few lines of JS running on the browser as possible. This is not very pragmatic and not always possible. As is, Reshaped is a great library and useful for many projects. The ideas in this issue may not be ideal for Reshaped.

**Additional context**

**Comment by blvdmitry** [comment]
Hey, this is a good call out. Some of the items on the list are definitely closer to reality than the others and I think we just need to go through them one by one. I think for the most of it – it's about the balance. Probably won't turn into a pure css solution due to additional features like handling selection transition in tabs or handling edge cases for the focus trapping in modals. That still means we could ship fewer javascript, so I'll look into this list and see how much each item saves us in the end.

Right now we're at 45kb js and 17kb of css for the whole set of components, while also supporting tree-shaking and not requiring any additional runtime dependencies like emotion or other css-in-js runtime like in many other libraries. Additionally we provide RSC support wherever possible which reduces the js shipped to the client in many cases.

**Comment by tatemz** [comment]
I recently ran across a use-case that touches on this issue. I am wanting to capture form fields using a modal UX without losing their data. I am curious if there is "Reshaped Way" of accomplishing this user experience. If so, I wonder if the "Reshaped Way" is closely aligned with an in browser way of doing things. 

Below is a code sandbox that shows the experience. I could also wire up the vanilla browser functionality with `<dialog>` elements and forms ([see docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog#usage_notes)).

https://codesandbox.io/p/sandbox/brave-browser-v798pf

**Comment by petewins** [comment]
@tatemz , I believe the inputs inside the Modal are not captured by the form when you submit even though your Modal is inside a `<form>` element because

1. The input fields inside that modal are only rendered in the dom when it is active. When the modal is closed the components inside are unmounted.
2. The Modal seems to portal the components elsewhere, so even if the input fields are persisted it might not be a child of the <form> element anymore. 

Perhaps there can be an option on the Modal to not portal and use the dialog?

**Comment by tatemz** [comment]
You are correct @petewins. Compare this to native browser functionality where DOM elements can be shown/hidden without being appended/destroyed

I have updated the code sandbox with a `<dialog>` example as well

**Comment by tatemz** [comment]
I have updated the code sandbox to handle `onChangeCommit` functionality
